encoding/kzg/prover/parametrized_prover.go

```diff
@@ -46,228 +69,118 @@ func (g *ParametrizedProver) EncodeBytes(inputBytes []byte) (*bn254.G1Affine, *b
-	}
-	close(jobChan)
-
-	// return last error
-	var err error
-	for w := uint64(0); w < numWorker; w++ {
-		wr := <-results
-		if wr.err != nil {
-			err = wr.err
+	// compute commit for the full poly
+	go func() {
+		start := time.Now()
+		commit, err := g.Computer.ComputeCommitment(inputFr)
+		commitmentChan <- CommitmentResult{
+			Commitment: *commit,
+			Err:        err,
+			Duration:   time.Since(start),
 		}
-	}
-
-	if err != nil {
-		return nil, fmt.Errorf("proof worker error: %v", err)
-	}
-
-	t0 := time.Now()
-
-	// compute proof by multi scaler multiplication
-	msmErrors := make(chan error, dimE*2)
-	for i := uint64(0); i < dimE*2; i++ {
-
-		go func(k uint64) {
-			_, err := sumVec[k].MultiExp(p.FFTPointsT[k], coeffStore[k], ecc.MultiExpConfig{})
-			// handle error
-			msmErrors <- err
-		}(i)
-	}
-
-	for i := uint64(0); i < dimE*2; i++ {
-		err := <-msmErrors
-		if err != nil {
-			fmt.Println("Error. MSM while adding points", err)
-			return nil, err
+	}()
+
+	go func() {
+		start := time.Now()
+		lengthCommitment, err := g.Computer.ComputeLengthCommitment(inputFr)
+		lengthCommitmentChan <- LengthCommitmentResult{
+			LengthCommitment: *lengthCommitment,
+			Err:              err,
+			Duration:         time.Since(start),
+		}
+	}()
+
+	go func() {
+		start := time.Now()
+		lengthProof, err := g.Computer.ComputeLengthProof(inputFr)
+		lengthProofChan <- LengthProofResult{
+			LengthProof: *lengthProof,
+			Err:         err,
+			Duration:    time.Since(start),
+		}
+	}()
+
+	go func() {
+		start := time.Now()
+		// compute proofs
+		paddedCoeffs := make([]fr.Element, g.NumEvaluations())
+		// polyCoeffs has less points than paddedCoeffs in general due to erasure redundancy
+		copy(paddedCoeffs, inputFr)
+
+		numBlob := 1
+		flatpaddedCoeffs := make([]fr.Element, 0, numBlob*len(paddedCoeffs))
+		for i := 0; i < numBlob; i++ {
+			flatpaddedCoeffs = append(flatpaddedCoeffs, paddedCoeffs...)
 		}
-	}
-
-	t1 := time.Now()
 
-	// only 1 ifft is needed
-	sumVecInv, err := p.Fs.FFTG1(sumVec, true)
-	if err != nil {
-		return nil, fmt.Errorf("fft error: %v", err)
-	}
+		proofs, err := g.Computer.ComputeMultiFrameProof(flatpaddedCoeffs, g.NumChunks, g.ChunkLength, g.NumWorker)
+		proofChan <- ProofsResult{
+			Proofs:   proofs,
+			Err:      err,
+			Duration: time.Since(start),
+		}
+	}()
 
-	t2 := time.Now()
+	lengthProofResult := <-lengthProofChan
+	lengthCommitmentResult := <-lengthCommitmentChan
+	commitmentResult := <-commitmentChan
+	rsResult := <-rsChan
+	proofsResult := <-proofChan
 
-	// outputs is out of order - buttefly
-	proofs, err := p.Fs.FFTG1(sumVecInv[:dimE], false)
-	if err != nil {
-		return nil, err
+	if lengthProofResult.Err != nil || lengthCommitmentResult.Err != nil ||
+		commitmentResult.Err != nil || rsResult.Err != nil ||
+		proofsResult.Err != nil {
+		return nil, nil, nil, nil, nil, multierror.Append(lengthProofResult.Err, lengthCommitmentResult.Err, commitmentResult.Err, rsResult.Err, proofsResult.Err)
```

Comment:
I think using https://pkg.go.dev/golang.org/x/sync/errgroup could be very idiomatic here.
I've never used errgroup myself but I think the idea is that here if one of the computation errors early, your current code will still wait for everything to finish, and then return the joint error.
An errgroup would short-circuit the other computations as soon as one goroutine errors out, and return that. Unless you really want to know all errors that happened (and not just the first one), then this could be useful to save resources?